Refactor dev docs (#4794)
strimzi/strimzi-kafka-operator: development-docs/DEV_GUIDE.md

more details.

## Build Pre-Requisites

### Command line tools

To build this project you must first install several command line utilities and a Kubernetes or OpenShift cluster.

- [`make`](https://www.gnu.org/software/make/) - Make build system
- [`mvn`](https://maven.apache.org/index.html) (version 3.5 and above) - Maven CLI
- [`helm`](https://helm.sh/) (version 3) - Helm Package Management System for Kubernetes
- [`asciidoctor` and `asciidoctor-pdf`](https://asciidoctor.org/) - Documentation generation.
    - Use `gem` to install the latest version for your platform.
- [`yq`](https://github.com/mikefarah/yq) - (version 4.2.1 and above) YAML manipulation tool.
    - **Warning:** There are several different `yq` YAML projects in the wild.
      Use [this one](https://github.com/mikefarah/yq). You need version 4 or above.
- [`docker`](https://docs.docker.com/install/) - Docker command line client
- [`shellcheck`](https://github.com/koalaman/shellcheck) - ShellCheck is a GPLv3 tool that gives warnings and
  suggestions for bash/sh shell scripts.

In order to use `make` these all need to be available on your `$PATH`.

### Mac OS

> The `make` build is using GNU versions of `find` and `sed` utilities and is not compatible with the BSD versions

Reviewer: ```suggestion
The `make` build is using GNU versions of `find`, `sed` and other utilities and is not compatible with the BSD versions
```

The tests run during the build can be skipped by setting the `MVN_ARGS` environment variable and passing that to the
make command:

        make clean
        make MVN_ARGS='-DskipTests -DskipITs' all

4. To use the newly built images, update the `install/cluster-operator/060-Deployment-strimzi-cluster-operator.yml` to obtain the images from your repositories on Docker Registry rather than the official Strimzi images:

    ```
    sed -Ei -e "s#(image|value): quay.io/strimzi/([a-z0-9-]+):latest#\1: $DOCKER_REGISTRY/$DOCKER_ORG/\2:latest#" \
4. To use the newly built images, update
   the `packaging/install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml` to obtain the images from your
   repositories on Docker Registry rather than the official Strimzi images:
   \
   \
   **Linux**
   ```
   sed -Ei -e "s#(image|value): quay.io/strimzi/([a-z0-9-]+):latest#\1: $DOCKER_REGISTRY/$DOCKER_ORG/\2:latest#" \
            -e "s#(image|value): quay.io/strimzi/([a-zA-Z0-9-]+:[0-9.]+)#\1: $DOCKER_REGISTRY/$DOCKER_ORG/\2#" \
            -e "s#([0-9.]+)=quay.io/strimzi/([a-zA-Z0-9-]+:[a-zA-Z0-9.-]+-kafka-[0-9.]+)#\1=$DOCKER_REGISTRY/$DOCKER_ORG/\2#" \

>             install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml

Reviewer 2: Without this line the commmand won't work?

Author: Good catch @tomncooper, will fix it up in my NZ morning :)

Reviewer: But it will need to be updated to point to `packaging/`.

5. The installation files assume you're installing into the namespace `myproject`. If you wish to use a different one,
   you'll need to replace it in the installation files.
   \
   \
   **Linux**

>        sed -Ei "s/myproject/<desired_namespace>/g" packaging/install/cluster-operator/*.yaml

Author: This should be consistent with the OS X example - using `*RoleBinding*.yaml` instead of `*.yaml`

>    **OS X**

Reviewer: Technically, it is macOS. So the original Mac OS was closer to the real name or?

Author: I didn't realise that they'd rebranded, pushed a change for all occurrences of OS X -> macOS. :+1:

can) and create a new folder named `JavaVirtualMachines` followed by creating a file named `openjdk-11.jdk`. The folder
structure after everything is said and done should look like `/Library/Java/JavaVirtualMachines/openjdk-11.jdk`. After
doing that run the command at the beginning again and this should link the file and allow you to use maven with OpenJDK
version 11.

### Kubernetes or OpenShift Cluster

In order to run the integration tests and test any changes made to the operators you will need a functioning Kubernetes
or OpenShift cluster. This can be a remote cluster or a local development cluster.

#### Minikube

The default Minikube setup should allow the integration tests to be run without additional configuration changes.

Sometimes however, updates to Minikube may prevent the test cluster context being correctly set. If you have errors due
to `oc` commands being called on your Minikube cluster then explicitly set the `TEST_CLUSTER` environment variable
before running the `make` commands.

    export TEST_CLUSTER=minikube

## Build and deploy from source

> To build Strimzi from a source the operator and Kafka code needs to be compiled into Docker container images and placed

Reviewer 3: ```suggestion
To build Strimzi from a source the operator and Kafka code needs to be compiled into container images and placed
```

in a location accessible to the Kubernetes/OpenShift nodes. The easiest way to make your personal Strimzi builds

> accessible, is to place them on the [Docker Hub](https://hub.docker.com/). The instructions below use this method,

Reviewer 3: ```suggestion
accessible, is to place them on the public container registry like [Docker Hub](https://hub.docker.com/) or quay (link). The instructions below use Docker Hub,
```

Author: Thanks @Frawless have pushed a new commit incorporating all your suggestions, plus a couple of other slight issues I found.

    make DOCKER_ORG=strimzi docker_tag

This labels your latest container build as `strimzi/operator:latest` and you can then deploy the standard CRDs without
changing the image targets. However, this will only work if all instances of the `imagePullPolicy:` setting are set
to `IfNotPresent` or `Never`. If not, then the cluster nodes will go to the upstream registry (Docker Hub by default)
and pull the official images instead of using your freshly built image.

## Helm Chart

The `strimzi-kafka-operator` Helm Chart can be installed directly from its source.

>     `helm install ./helm-charts/helm3/strimzi-kafka-operator`

Reviewer: Should this point to `packaging/`?
